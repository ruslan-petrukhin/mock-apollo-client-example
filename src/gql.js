import gql from 'graphql-tag';

export const GET_PROCESS = gql`
  query getProcess {
    process @client {
      flow
    }
  }
`;

export const SAVE_USER = gql`
  mutation saveUser($fullName: String!, $email: String!) {
    saveUser(fullName: $fullName, email: $email) @client
  }
`;

export const GET_USER = gql`
  query getUser {
    userDetails @client {
      fullName
      email
    }
  }
`;

export const getResolvers = (cache) => ({
  Mutation: {
    saveUser(_root, {fullName, email}) {
      cache.writeData({
        data: {
          userDetails: {__typename: 'UserDetails', fullName, email},
        },
      });
    },
  },
});
