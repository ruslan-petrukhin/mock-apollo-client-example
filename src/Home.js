import React, {useCallback, useState, useMemo} from 'react';
import {
  SafeAreaView,
  StyleSheet,
  ScrollView,
  Text,
  TextInput,
  TouchableOpacity,
} from 'react-native';
import {useQuery, useMutation} from '@apollo/react-hooks';

import {GET_PROCESS, SAVE_USER} from './gql';

const Home = () => {
  const {data} = useQuery(GET_PROCESS);
  const [saveUser] = useMutation(SAVE_USER);

  const [fullName, setFullName] = useState('');
  const [email, setEmail] = useState('');

  const disabled = useMemo(() => {
    if (!fullName || !email) {
      return true;
    } else {
      return false;
    }
  }, [fullName, email]);

  const handleChangeFullName = useCallback((text) => {
    setFullName(text);
  }, []);

  const handleChangeEmail = useCallback((text) => {
    setEmail(text);
  }, []);

  const submit = useCallback(async () => {
    await saveUser({
      variables: {
        fullName,
        email,
      },
    });
    if (data.process.flow === 'new') {
      console.log('navigate to first screen');
    } else {
      console.log('navigate to second screen');
    }
  }, [data, saveUser, fullName, email]);

  return (
    <SafeAreaView style={styles.base}>
      <ScrollView
        style={styles.base}
        contentContainerStyle={styles.content}
        alwaysBounceVertical={false}>
        <Text>FullName</Text>
        <TextInput onChangeText={handleChangeFullName} style={styles.input} />
        <Text>Email</Text>
        <TextInput onChangeText={handleChangeEmail} style={styles.input} />
        <TouchableOpacity
          onPress={submit}
          style={[styles.submit, disabled && styles.disabled]}
          disabled={disabled}>
          <Text style={styles.submitText}>Submit</Text>
        </TouchableOpacity>
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  base: {
    flex: 1,
  },
  content: {
    flexGrow: 1,
    justifyContent: 'center',
    paddingHorizontal: 48,
  },
  input: {
    padding: 8,
    marginTop: 4,
    marginBottom: 24,
    borderWidth: 1,
    borderColor: '#666666',
  },
  submit: {
    height: 40,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'green',
  },
  submitText: {
    fontSize: 16,
    lineHeight: 19,
    color: 'white',
  },
  disabled: {
    opacity: 0.2,
  },
});

export default Home;
