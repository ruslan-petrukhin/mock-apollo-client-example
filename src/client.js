import {ApolloClient} from 'apollo-client';
import {createHttpLink} from 'apollo-link-http';
import {InMemoryCache} from 'apollo-cache-inmemory';

import {getResolvers} from './gql';

const link = createHttpLink({uri: ''});
const cache = new InMemoryCache();

export const client = new ApolloClient({
  link,
  cache,
  resolvers: getResolvers(cache),
});

const data = {
  process: {
    __typename: 'Process',
    flow: 'new',
  },
  userDetails: {
    __typename: 'UserDetails',
    fullName: '',
    email: '',
  },
};

cache.writeData({data});

export default client;
