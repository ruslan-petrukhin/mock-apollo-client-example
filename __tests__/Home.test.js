import 'react-native';
import React from 'react';
import {ApolloProvider} from '@apollo/react-hooks';
import {createMockClient} from 'mock-apollo-client';
import renderer from 'react-test-renderer';

import Home from '../src/Home';
import {GET_PROCESS, SAVE_USER} from '../src/gql';

const wait = async (time) => {
  await new Promise((resolve) => setTimeout(resolve, time));
};

const mockSaveUser = jest.fn().mockResolvedValue({
  data: {saveUser: {}},
});

describe('Home', () => {
  let wrapper;

  const mockClient = createMockClient();
  mockClient.setRequestHandler(GET_PROCESS, () =>
    Promise.resolve({data: {process: {flow: 'new'}}}),
  );
  mockClient.setRequestHandler(SAVE_USER, mockSaveUser);

  beforeEach(async () => {
    await renderer.act(() => {
      wrapper = renderer.create(
        <ApolloProvider client={mockClient}>
          <Home />
        </ApolloProvider>,
      );
      return wait(0);
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', async () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should save user and navigate to first screen on form submit', async () => {
    jest.spyOn(console, 'log').mockImplementation(jest.fn());

    await renderer.act(async () => {
      wrapper.root.findAllByType('TextInput')[0].props.onChangeText('John Doe');
      await wait(0);

      wrapper.root
        .findAllByType('TextInput')[1]
        .props.onChangeText('john_doe@email.com');
      await wait(0);

      wrapper.root.findAllByType('View')[1].props.onClick();
      return wait(0);
    });

    expect(mockSaveUser).toHaveBeenCalledWith({
      fullName: 'John Doe',
      email: 'john_doe@email.com',
    });
    expect(console.log).toHaveBeenCalledWith('navigate to first screen');
  });
});
