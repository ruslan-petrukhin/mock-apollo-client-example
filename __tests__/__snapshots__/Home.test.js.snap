// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    alwaysBounceVertical={false}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "justifyContent": "center",
        "paddingHorizontal": 48,
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <Text>
        FullName
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "#666666",
            "borderWidth": 1,
            "marginBottom": 24,
            "marginTop": 4,
            "padding": 8,
          }
        }
        underlineColorAndroid="transparent"
      />
      <Text>
        Email
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "#666666",
            "borderWidth": 1,
            "marginBottom": 24,
            "marginTop": 4,
            "padding": 8,
          }
        }
        underlineColorAndroid="transparent"
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "green",
            "height": 40,
            "justifyContent": "center",
            "opacity": 0.2,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 16,
              "lineHeight": 19,
            }
          }
        >
          Submit
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
